README.md
# pimatic-i2c 

Pimatic plugin for the I2C based sensors

This plugin makes I2C based sensors available in Pimatic. Devices are connected via the I2C port of your computer.
The plugin is tested on a raspberry pi.

The plugin can be installed via the plugins page of Pimatic.

## Config of the plugin
```
{
  debug: "Debug mode. Writes debug messages to the Pimatic log, if set to true."
}
```
After installation of the plugin (and restart) devices can be added via the devices pages in Pimatic.

## The Ina219 Device

![](assets/ina219.png)  


Device configuraton:

```
{
 device: "Device number to use (prefix /dev/i2c- is automatically added)"
   default: 1
 address: "Address of the sensor"
   enum: [0x40,0x41,0x44,0x45]
   default: 0x40
 interval: "Sensor read interval in ms"
   default: 10000
}
```

The Ina219 provides 2 attributes
- Voltage (V)
- Current (A)

The attributes are updated and visible in the Gui every \<interval\> milliseconds.

## The Mcp3424 Device

Device configuraton:

```
{
 device: "Device number to use (prefix /dev/i2c- is automatically added)"
   	default: 1
 address: "Address of the sensor"
   	enum: [0x40,0x41,0x44,0x45]
   	default: 0x40
 forwardPowerChannel: "Channel of the forwardPower sensor (1..4)"
 	default: 1
 reflectedPowerChannel: "Channel of the reflectedPower sensor (1..4)"
 	default: 2
 gain: ""
 	default: 1
 resolution: ""
 	default: 1
 interval: "Sensor read interval in ms"
   default: 10000
}
```

The Mcp3424 provides 2 attributes
- forwardPower (W)
- reflectedPower (W)

The attributes are updated and visible in the Gui every \<interval\> milliseconds.

---
This plugin is based on the Ina219 library from [brettmarl](https://github.com/brettmarl/node-ina219#readme).

The minimum node requirement for this plugin is node v8!


HISTORY.md
# Release History

* 20210101, v0.0.1
	* initial release
* 20210128, v0.1.0
	* initial release
* 20210130, v0.1.2
	* edit
* 20210130, v0.1.3
	* change in reading channels
* 20210201, v0.1.4
	* update register read
* 20210202, v0.1.5
	* correct address handling
* 20210203, v0.1.6
	* update channel handling
* 20210213, v0.1.7
	* added offset
* 20210213, v0.1.10
	* edit address hexadecimal handling
	* no default address
* 20210222, v0.1.11
	* trunk display value to 0


adapters/mcp3424.js
"use strict";
/*
 * Node driver for MCP3424
 */
var i2c = require('i2c-bus');  // https://github.com/fivdi/i2c-bus

/**
 * Mcp3424 is the main class exported from the Node module
 *
 */
var MCP3424 = function() {}

/**
 * Callback for standard oncomplete
 *
 * @callback onCompleteCallback
 */

/**
 * Callback for returning a single value
 *
 * @callback onHaveValueCallback
 * @param {int} value - value returned by async operation
 */

// ===========================================================================
//   I2C Values
// ==========================================================================

var MCP342X_ADDRESS         = 0x68  ; // default address
// ===========================================================================
var MCP342X_RES_SHIFT       = 2  ;

var MCP342X_GAIN_FIELD      = 0x03  ; // Gain field
var MCP342X_GAIN_X1         = 0x00  ; // Gain 1
var MCP342X_GAIN_X2         = 0x01  ; // Gain 2
var MCP342X_GAIN_X4         = 0x02  ; // Gain 4
var MCP342X_GAIN_X8         = 0x03  ; // Gain 8

var MCP342X_RES_FIELD       = 0x0C  ; // resolution field
var MCP342X_12_BIT          = 0x00  ; // 12 bit Resolution Mask
var MCP342X_14_BIT          = 0x04  ; // 14 bit Resolution Mask
var MCP342X_16_BIT          = 0x08  ; // 16 bit Resolution Mask
var MCP342X_18_BIT          = 0x0C  ; // 18 bit Resolution Mask

var MCP342X_CONTINUOUS      = 0x10  ;

var MCP342X_CHAN_FIELD      = 0x60  ; // channel field
var MCP342X_CHANNEL_1       = 0x00  ; // channel 1
var MCP342X_CHANNEL_2       = 0x20  ; // channel 2
var MCP342X_CHANNEL_3       = 0x40  ; // channel 3
var MCP342X_CHANNEL_4       = 0x60  ; // channel 4

var MCP342X_START           = 0x80  ; // Operating Mode Mask
var MCP342X_BUSY            = 0x80;
//var MCP3424 ;



/**
  * Called to initilize the MCP3424.
  * @param {string} address - Address you want to use. Defaults to MCP342X_ADDRESS
  * @param {integer} gain
  * @param {integer} resolution
  * @param {integer} busNumber - the number of the I2C bus/adapter to open, 0 for /dev/i2c-0, 1 for /dev/i2c-1, (See github.com/fivdi/i2c-bus)
  */
MCP3424.prototype.init = function (address, gain, resolution, busNumber) {

  // defaults
  address = typeof address !== 'undefined' ? address : MCP342X_ADDRESS;
  busNumber = typeof busNumber !== 'undefined' ? busNumber : 1;

  this.log("init:: " + address + " | " + gain + " | " + resolution + " | " + busNumber)
  this.resolution = resolution;
  this.gain = gain;
  this.channel = [];
  this.currChannel = 0;
  this.address = address;

  this.wire = i2c.openSync(busNumber);
  var devices = this.wire.scanSync();
  if (address in devices) {
    this._readDataContiuously();    
  } else {
    this.log("init:: device address '" + address + "' not available")
  }
}

/**
  * Enabled debug logging to console.log
  * @param {bool} enable - True to enable, False to disable
  */
MCP3424.prototype.enableLogging  = function (enable) {

  this.loggingEnabled = enable;
}

/**
  * Logs a string to the console if logging enabled
  * @param {string} s - String to log
  */
MCP3424.prototype.log  = function (s) {

    if (this.loggingEnabled)
        console.log(s);
}

/**
  * Writes a 16 bit value over I2C
  * @param {integer} register - Register to read from (One of INA219_REG_*)
  * @param {integer} value - Value to be written
  * @param {writeRegisterCallback} callback - Callback to be invoked when complete
  */
/*
MCP3424.prototype.writeRegister  = function (register, value, callback) {

    var bytes = Buffer.alloc(2);

    bytes[0] = (value >> 8) & 0xFF;
    bytes[1] = value & 0xFF

    this.wire.writeI2cBlockSync(this.address, register, 2, bytes);
    callback(null);
}
*/
/**
  * Reads a 16 bit value over I2C
  * @param {integer} register - Register to read from (One of INA219_REG_*)
  * @param {onHaveValueCallback} callback - Callback to be invoked when complete
  */

/*
MCP3424.prototype.readRegister  = function (register, callback) {

    var res = Buffer.alloc(2);

    this.wire.readI2cBlockSync(this.address, register, 2, res);

    var value = res.readInt16BE();

    this.log("::readRegister => [" + res[0] + ", " + res[1] + "]");

    callback(value);
}

MCP3424.prototype.getBusVoltage_raw  = function (callback) {

    this.log("getBusVoltage_raw");

    var $this = this;
    this.readRegister(INA219_REG_BUSVOLTAGE, function (value) {

        $this.log("getBusVoltage_raw RET: " + value);

        //  Shift to the right 3 to drop CNVR and OVF and multiply by LSB
        callback( (value >> 3) * 4);

    });
}

MCP3424.prototype.getVoltage = function(chan) {
    return ((this.getMv(chan) * (0.0005 / this._getPga())) * 2.471)
*/

MCP3424.prototype.getMv = function(chan) {
    return this.channel[chan];
}

MCP3424.prototype.getVoltage = function(chan) {
    return ((this.getMv(chan) * (0.0005 / this._getPga())) * 2.471)
}

MCP3424.prototype._getMvDivisor = function() {
    var mvDivisor;
    return mvDivisor = 1 << (this.gain + 2 * this.res);
}

MCP3424.prototype._getAdcConfig = function(chan) {
    var adcConfig;
    adcConfig = MCP342X_CHANNEL_1 | MCP342X_CONTINUOUS;
    return adcConfig |= chan << 5 | this.resolution << 2 | this.gain;
}

MCP3424.prototype._changeChannel = function(chan) {
    var command, wbuf;
    command = this._getAdcConfig(chan);
    wbuf = Buffer.alloc(1)
    return this.wire.i2cWrite(command, 1, wbuf, function(err, bytesWritten, buffer) {
        if (err !== null) {
            return console.log(err);
        }
    });
}

MCP3424.prototype._readDataContiuously = function() {
    var self;
    self = this;
    return setInterval((function() {
        self._readData(self.currChannel);
    }), 5000); //was 10
}

MCP3424.prototype._nextChannel = function() {
    this.currChannel++;
    if (this.currChannel === 4) {
        return this.currChannel = 0;
    }
}

MCP3424.prototype._readData = function(chan) {
    var adcConfig, result, self, statusByte, rbuf;
    self = this;
    adcConfig = this._getAdcConfig(chan);
    result = 0;
    statusByte = 0;
    rbuf = Buffer.alloc(4);
    //bus.i2cRead(addr, length, buffer, cb)
    return this.wire.i2cRead(adcConfig, 4, rbuf, function(err, bytesRead, res) {
        var byte0, byte1, byte2;
        if (err !== null) {
            console.log(err);
        }
        if ((adcConfig & MCP342X_RES_FIELD) === MCP342X_18_BIT) {
            byte0 = res[0];
            byte1 = res[1];
            byte2 = res[2];
            statusByte = res[3];
            result = byte2 | byte1 << 8 | byte0 << 16;
        } else {
            byte0 = res[0];
            byte1 = res[1];
            statusByte = res[2];
            result = byte1 | byte0 << 8;
        }
        if ((statusByte & MCP342X_BUSY) === 0) {
            self.channel[self.currChannel] = result / self._getMvDivisor();
            self._nextChannel();
            return self._changeChannel(self.currChannel);
        } else {
            return "err";
        }
    });
}

MCP3424.prototype._getPga = function() {
    var gain = this.gain;

    if (gain == MCP342X_GAIN_X1) {
        return 0.5;
    }

    if (gain == MCP342X_GAIN_X2) {
        return 1;
    }

    if (gain == MCP342X_GAIN_X4) {
        return 2;
    }

    if (gain == MCP342X_GAIN_X8) {
        return 4;
    }
}


// export is a Singleton
module.exports = new MCP3424();


adapters/ReadADC_i2c.js
/****************************************************************
* Purpose: Proof of concept on how to read values from and configure the mcp3424
*          The code is targeted for a raspberry pi expansionboard that 
*          has 8 ports with a voltage divider allowing measuring of 0-14V
*
* Copyright: Peter Sjöberg
*
* License: GPL3
*
* Note: i2c-bus is required, can be installed with
*           npm install i2c-bus
*
* 2019-07-16  Peter Sjoberg <peters-src AT techwiz DOT ca>
*   Created as a partial port of Read_mcp3425.py
*
*/


'use strict';

/* ports are numbered from 1-32. The each board has two chips and each chip has four channels.
 * board 1 has
 *  port1  = chip address[1] channel 1
 *  port2  = chip address[1] channel 2
 *  port3  = chip address[1] channel 3
 *  port4  = chip address[1] channel 4
 *  port5  = chip address[2] channel 1
 *  port6  = chip address[2] channel 2
 *  port7  = chip address[2] channel 3
 *  port8  = chip address[2] channel 4
 * board 2 continues in same pattern
 *  port9  = chip address[3] channel 1
 *  port10 = chip address[3] channel 2
 * and so on
 */

var MCP3424 = function() {}

// default values
const BITS=18
const GAIN=1

const MCP3424_RDY = 0x80

const MCP3424_Channel = {
    1 : 0x00,
    2 : 0x20,
    3 : 0x40,
    4 : 0x60,
    "mask" : 0x60 }

const MCP3424_OC = 0x10

// 12bit / 1mV       /  240 SPS / -2048   --   2047
// 14bit / 250 uV    /   60 SPS / -8192   --   8191
// 16bit / 62.5uV    /   15 SPS / -32768  --  32767
// 18bit / 15.625uV  / 3.75 SPS / -131072 -- 131071
const MCP3424_bits = {
    12 : 0x00,
    14 : 0x04,
    16 : 0x08,
    18 : 0x0c,
    "mask" : 0x0c}

const MCP3424_GAIN = {
    1 : 0x00,   // 2.048V
    2 : 0x01,   // 1.024V
    4 : 0x02,   // 0.512V
    8 : 0x03,   // 0.256V
    "mask" : 0x03}

// Addresses of the ADC we want to access
var addresses=[
    0x68,0x69,
    0x6a,0x6b,
    0x6c,0x6d,
    0x6e,0x6f
]

const R1=120000;
const R2=20000;
const maxport=4; // single mcp3424

const i2c = require('i2c-bus');
const i2c1 = i2c.openSync(1); // 1 for i2c bus 1

// The following formula/constants was created by calculate the proper Rl value over several voltage
// samples and then reverse engineer the formula and constants needed to create the correct ADC load.
/***********************/
var TrimConstant

TrimConstant=[]
TrimConstant[12]=[]
TrimConstant[12][1]=[]
TrimConstant[12][1]['Const1']=-7.29644998741772e+18
TrimConstant[12][1]['Const2']=1.8252066801568475e+19
TrimConstant[12][1]['Const3']=-9.561470143248126e+18
TrimConstant[12][2]=[]
TrimConstant[12][2]['Const1']=-1.4898381548524163e+19
TrimConstant[12][2]['Const2']=1.8452580931431174e+19
TrimConstant[12][2]['Const3']=-4.805276209813626e+18
TrimConstant[12][4]=[]
TrimConstant[12][4]['Const1']=-10118072.271809563
TrimConstant[12][4]['Const2']=6797352.007977883
TrimConstant[12][4]['Const3']=-263832.86114829977
TrimConstant[12][8]=[]
TrimConstant[12][8]['Const1']=-22986374.244420122
TrimConstant[12][8]['Const2']=7536052.779157299
TrimConstant[12][8]['Const3']=-258788.32091889498
TrimConstant[14]=[]
TrimConstant[14][1]=[]
TrimConstant[14][1]['Const1']=-547128.0816393106
TrimConstant[14][1]['Const2']=1959171.5248400136
TrimConstant[14][1]['Const3']=413393.9108785477
TrimConstant[14][2]=[]
TrimConstant[14][2]['Const1']=-1303354.7551292207
TrimConstant[14][2]['Const2']=2216492.26594754
TrimConstant[14][2]['Const3']=252631.72247966885
TrimConstant[14][4]=[]
TrimConstant[14][4]['Const1']=-603659.535929856
TrimConstant[14][4]['Const2']=1058285.263214671
TrimConstant[14][4]['Const3']=238635.61023626427
TrimConstant[14][8]=[]
TrimConstant[14][8]['Const1']=-321542.9453517512
TrimConstant[14][8]['Const2']=506255.5322955709
TrimConstant[14][8]['Const3']=174575.86010087654
TrimConstant[16]=[]
TrimConstant[16][1]=[]
TrimConstant[16][1]['Const1']=-538886.6327368342
TrimConstant[16][1]['Const2']=1929058.1036710313
TrimConstant[16][1]['Const3']=418277.36598391004
TrimConstant[16][2]=[]
TrimConstant[16][2]['Const1']=-1093021.2089518933
TrimConstant[16][2]['Const2']=2078583.7929054706
TrimConstant[16][2]['Const3']=252687.99370679908
TrimConstant[16][4]=[]
TrimConstant[16][4]['Const1']=-1025665.3681316198
TrimConstant[16][4]['Const2']=1238120.803775412
TrimConstant[16][4]['Const3']=229019.1443321319
TrimConstant[16][8]=[]
TrimConstant[16][8]['Const1']=-354280.916034503
TrimConstant[16][8]['Const2']=517745.8280310895
TrimConstant[16][8]['Const3']=175604.3187191613
TrimConstant[18]=[]
TrimConstant[18][1]=[]
TrimConstant[18][1]['Const1']=-549702.5567547062
TrimConstant[18][1]['Const2']=1957717.4623683672
TrimConstant[18][1]['Const3']=387836.79665354657
TrimConstant[18][2]=[]
TrimConstant[18][2]['Const1']=-1361374.6457417163
TrimConstant[18][2]['Const2']=2135565.100213241
TrimConstant[18][2]['Const3']=255336.17805034263
TrimConstant[18][4]=[]
TrimConstant[18][4]['Const1']=-1245462.6278009678
TrimConstant[18][4]['Const2']=1159128.6214418625
TrimConstant[18][4]['Const3']=241634.35669704582
TrimConstant[18][8]=[]
TrimConstant[18][8]['Const1']=231079.04899498075
TrimConstant[18][8]['Const2']=389831.38249999925
TrimConstant[18][8]['Const3']=182839.62771505647

//variable array that holds all values
var data=[];

// check on https://www.codota.com/code/javascript/modules/i2c-bus
MCP3424.prototype.readADC = function (port){
    var size
    var i2cdata=Buffer(4); // place to store values read from the ADC

    //console.log("data[port]['bits']: "+data[port]['chipaddr'])

    if (data[port]['bits'] == 18){
        size=4;
    }else{
        size=3;
    }
    try {
        data[port]['valid']=true;
        //console.log("port "+port+", cmd: 0x"+(MCP3424_RDY | MCP3424_OC | MCP3424_Channel[data[port]['channel']] |MCP3424_bits[data[port]['bits']] | MCP3424_GAIN[data[port]["gain"]]).toString(16))
        i2c1.readI2cBlockSync(data[port]['chipaddr'], MCP3424_RDY | MCP3424_OC | MCP3424_Channel[data[port]['channel']] |MCP3424_bits[data[port]['bits']] | MCP3424_GAIN[data[port]["gain"]], size,i2cdata);
    }

    catch(e){ // 
        //console.log("read port no "+port+" failed, marking it invalid");
        //console.log(e);
        data[port]['valid']=false;
    }
    data[port]['raw']=i2cdata;
    data[port]['status']=i2cdata[size-1]
    //console.log("    status for port "+port+" = "+data[port]['status'])
    return data[port]['status']
} // readADC

/****************
* convert the raw ADC value to a voltage
*/

MCP3424.prototype.adc2volt = function (port){
        
    let bits=data[port]['bits'];
    let val=data[port]['raw'];
    let gain=data[port]['gain'];
    
    let LSB=(2*2.048)/(2**bits);
    let rawVal,Const1,Const2,Const3

    if (bits == 18){
        rawVal = (val[0] & 0x3) <<16 | val[1] << 8 | val[2];
    }else if (bits == 16){
        rawVal = val[0] <<8 | val[1];
    }else if (bits == 14){
        rawVal = (val[0] & 0x3f)<<8 | val[1];
    }else if (bits == 12){
        rawVal = (val[0] & 0xf)<<8 | val[1];
    }
    
    if (val[0] & 0x80){
        rawVal -= 2**bits;
    }
    data[port]['rawVal']=rawVal;
    data[port]['adcV']=(data[port]['rawVal']*LSB)/data[port]['gain'];
    data[port]['mV']=data[port]['adcV']/1000;

    // ADC load by the specsheet but doesn't work so good
    // The following formula/constants was created by calculate the proper Rl value over several voltage
    // samples and then reverse engineer the formula and constants needed to create the correct ADC load.

    //console.log("adcV port "+port+", "+data[port]['adcV']+", raw: "+rawVal)

    return
    
    try{
        Const1=TrimConstant[bits][gain]['Const1']
        Const2=TrimConstant[bits][gain]['Const2']
        Const3=TrimConstant[bits][gain]['Const3']
    }
    catch(e){
        //console.log("entering catch block");
        console.log(e);
        //console.log("leaving catch block");
        console.log("ERROR, unknown bits of "+bits+" or gain of "+gain+" - ABORT")
        console.log(TrimConstant[18]);
        console.log(" ---")
        console.log(TrimConstant[18][1]);
        process.exit(112);
    }

    // calculate ADC Internal load resistance
    data[port]['Rl']=Const1*Math.abs(data[port]['adcV'])**2+Const2*Math.abs(data[port]['adcV'])+Const3;

    if (data[port]['R1'] >0){
        data[port]['I']=data[port]['adcV']/data[port]['R2'];
        data[port]['volt']=data[port]['I']*(data[port]['R1']+data[port]['R2']);
        data[port]['trueR2']=1/(1/data[port]['R2']+1/data[port]['Rl']);
        data[port]['trueI']=(data[port]['adcV'])/data[port]['trueR2'];
        data[port]['trueV']=data[port]['trueI']*(data[port]['R1']+data[port]['trueR2']);
    }else{
        data[port]['I']=data[port]['adcV']/data[port]['Rl'];
        data[port]['volt']=data[port]['adcV'];
        data[port]['trueV']=data[port]['adcV'];
    }
    console.log("   port "+port+"  trueV: "+data[port]['trueV']);
} // adc2volt


MCP3424.prototype.setOpt = function (port,option,val){ //No error checking, just set the value...
    data[port][option]=val;
    if (option == "bits"){
        data[port]['LSB']=(2*2.048)/(2**data[port]['bits'])
    }
} // setOpt

/********************************
* Prime the data array with values
*/

MCP3424.prototype.setup = function (address, channels) {
    var board=1
    var chip=1
    var channel=1
    var port

    //Fill in values for all <maxport> ports, non existing ones get status of invalid (in readADC)
    for (var i = 0; i < channels.length; i++) {
        port = channels[i];
        data[port]= {}
        data[port]['bits']=BITS;
        data[port]['gain']=GAIN;
        data[port]['board']=board;
        data[port]['chip']=chip;
        data[port]['chipaddr']=address; //addresses[board*2+chip-3];
        data[port]['channel']=channel;
        //console.log("port "+port+", board "+board+", chip "+chip+", channel "+channel+", addr idx: "+(board*2+chip-3)+" is 0x"+parseInt(data[port]['chipaddr'], 10).toString(16)+", bits: "+data[port]['bits']);
        data[port]['R1']=R1; // upper resistor in voltage divider
        data[port]['R2']=R2; // lower resistor in voltage divider
        data[port]['mV']=-999999
        data[port]['volt']=-999999
        data[port]['LSB']=(2*2.048)/(2**BITS)
        data[port]['cnt']=0
        this.readADC(port)
        /* no need to wait for RDY, we mostly just check if the chip is there and prime the data array
        while (readADC(port) & MCP3424_RDY){
            data[port]['cnt']+=1
            if (data[port]['cnt']>1000){
                break
            }
        }
        */
        this.adc2volt(port); // to fill in rawVal and so on
        
        if (channel==4){
            channel=1;
            if (chip==2){
                chip=1;
                board+=1;
            }else{
                chip=2;
            }
        }else{
            channel+=1;
        }

    }
    //console.log("After setup: "+JSON.stringify(data,null,2));
} // setup

function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

MCP3424.prototype.readChannel = function (channel) {
    //var succes = this.readADC(channel)
    //await delay(1000)

    data[channel]['cnt']=0
    while (this.readADC(channel) & MCP3424_RDY){
        data[channel]['cnt']+=1
        //console.log("count: "+data[channel]['cnt'])
        if (data[channel]['cnt']>1000){
            return {succes: false}
            break
        }
    }
    this.adc2volt(channel)
    var volt = data[channel]['adcV']
    console.log("adcV channel "+channel+", "+volt)

    return {succes: true, adcV: volt}

    /*
    if (this.readADC(channel)) {
        //await delay(500)
        this.adc2volt(channel)
        var volt = data[channel]['adcV']
        console.log("adcV channel "+channel+", "+volt)

        return {succes: true, adcV: volt}
    } else {
        return {succes: false}
    }
    */
}

/****************
* small sample on how to use it
*/

/*
setup();
var start
var end

for (let loopcnt=0;loopcnt<2;loopcnt++){
    start = new Date();

    for (let port=1;port<=32;port++){
        setOpt(port,'gain',1);
        setOpt(port,'bits',18);
        if (data[port]['valid']==false){
            continue
        }

        data[port]['cnt']=0
        while (readADC(port) & MCP3424_RDY){
            data[port]['cnt']+=1
            if (data[port]['cnt']>1000){
                break
            }
        }
        adc2volt(port); // to fill in rawVal and so on
        console.log("Port "+port+": adcV="+data[port]['adcV']+", "+data[port]['trueV']+" Volt  ("+data[port]['cnt']+" cnt)");
        //console.log("Port "+port+": adcV="+data[port]['adcV']+", "+data[port]['trueV']+" Volt");
    }
    end = new Date() - start;
    console.log("Execution time: "+end+" ms");
    console.log(" ")
}
*/

module.exports = new MCP3424();
